### PR TITLE
added refill chest key option

### PR DESCRIPTION
**Pull request by JohnStyleZ**

### CODE REVIEW

File: src/bot.py [hunk 1]
```diff
@@ -273,10 +273,14 @@ def on_maintenance(self):
 
         # Check if we are out of tps or need repairing
         need_repair = self._ui_manager.repair_needed()
-        if self._tps_left < random.randint(3, 5) or need_repair or self._config.char["always_repair"]:
+        if self._tps_left < random.randint(3, 5) or need_repair or self._config.char["always_repair"] or self._config.char["always_refill_keys"] & self._config.routes["run_arcane"]:
```
Comment (aeon0) [inline src/bot.py:276]: I mean it works. But it is a bit weird to use an bitwise and for bool if condition. I think it is more readable if you use  "and" here.

File: src/bot.py [hunk 2]
```diff
@@ -273,10 +273,14 @@ def on_maintenance(self):
...
             if need_repair: Logger.info("Repair needed. Gear is about to break")
             else: Logger.info("Repairing and buying TPs at next Vendor")
-            self._curr_loc = self._town_manager.repair_and_fill_tps(self._curr_loc)
+            if self._config.char["always_refill_keys"] & self._config.routes["run_arcane"]:
```
Comment (aeon0) [inline src/bot.py:279]: keep code dry. you use twice the same condition (always_refill_keys and run_arcane). Store it in a variable similar to how it's done with need_repair. Something like `buy_keys`

File: src/bot.py [hunk 3]
```diff
@@ -273,10 +273,14 @@ def on_maintenance(self):
...
+                self._curr_loc = self._town_manager.refill_keys(self._curr_loc)
+                if not self._curr_loc:
+                	return self.trigger_or_stop("end_game", failed=True)
```
Comment (aeon0) [inline src/bot.py:282]: Is the indentation correct here?

File: src/bot.py [hunk 4]
```diff
@@ -273,10 +273,14 @@ def on_maintenance(self):
...
+            else: self._curr_loc = self._town_manager.repair_and_fill_tps(self._curr_loc)
```
Comment (aeon0) [inline src/bot.py:283]: Should this be in an else branch? Couldnt we need to refill keys AND might have to repair?

**Review by aeon0 — CHANGES_REQUESTED** [review]
Whats the rule for vendors selling keys? All vendors that repair also sell keys? Shouldnt we have one function that basically does all things: "repairs", "restocks tps" and, if selected in the configs, also buys keys?

**Comment by JohnStyleZ** [comment]
> Whats the rule for vendors selling keys? All vendors that repair also sell keys? Shouldnt we have one function that basically does all things: "repairs", "restocks tps" and, if selected in the configs, also buys keys?

the way it refill keys just like refilling tps, you have to have a key in your inventory in order for it to work, also since you are already at the vendor, i dont see why not just repair and refill tps too. the refill key only works if arcane and refill key are selected, cuz you dont need it for other runs.

**Comment by JohnStyleZ** [comment]
> Whats the rule for vendors selling keys? All vendors that repair also sell keys? Shouldnt we have one function that basically does all things: "repairs", "restocks tps" and, if selected in the configs, also buys keys?

the way it refill keys just like refilling tps, you have to have a key in your inventory in order for it to work, also since you are already at the vendor, i dont see why not just repair and refill tps too. the refill key only works if arcane and refill key are selected, cuz you dont need it for other runs.

File: src/bot.py (hunk 4 above)
Comment (JohnStyleZ) [reply src/bot.py:283]: It refill key also repair and refill tps

File: src/bot.py (hunk 3 above)
Comment (JohnStyleZ) [reply src/bot.py]: fixed

File: src/bot.py (hunk 4 above)
Comment (aeon0) [reply src/bot.py:283]: Thats my problem, cant we reuse the already implemented repair_and_tp to take a flag `refill_keys: bool = False`. It looks like pretty much duplicated code now in the town manager.

**Review by aeon0 — CHANGES_REQUESTED** [review]

File: src/bot.py (hunk 4 above)
Comment (JohnStyleZ) [reply src/bot.py:283]: since we are already at the vendor, why not repair and tp?

**Comment by mgleed** [comment]
Thank you for your work on this. Given aeon's advise to refactor in order to not have to copy `repair_and_tp` and no recent updates, will close for now. Feature is coming soon, though. #484 will purchase keys, pickup keys on the ground if needed, and will decide if you want to use keys by checking for them in inventory

**Pull request closed without merging**